api/apps/v1alpha1/nemo_evaluator_types.go

```diff
@@ -0,0 +1,764 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"fmt"
+	"maps"
+	"os"
+
+	rendertypes "github.com/NVIDIA/k8s-nim-operator/internal/render/types"
+	utils "github.com/NVIDIA/k8s-nim-operator/internal/utils"
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	autoscalingv2 "k8s.io/api/autoscaling/v2"
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	// NemoEvaluatorConditionReady indicates that the NEMO EvaluatorService is ready.
+	NemoEvaluatorConditionReady = "Ready"
+	// NemoEvaluatorConditionFailed indicates that the NEMO EvaluatorService has failed.
+	NemoEvaluatorConditionFailed = "Failed"
+
+	// NemoEvaluatorStatusPending indicates that NEMO EvaluatorService is in pending state
+	NemoEvaluatorStatusPending = "Pending"
+	// NemoEvaluatorStatusNotReady indicates that NEMO EvaluatorService is not ready
+	NemoEvaluatorStatusNotReady = "NotReady"
+	// NemoEvaluatorStatusReady indicates that NEMO EvaluatorService is ready
+	NemoEvaluatorStatusReady = "Ready"
+	// NemoEvaluatorStatusFailed indicates that NEMO EvaluatorService has failed
+	NemoEvaluatorStatusFailed = "Failed"
+)
+
+// NemoEvaluatorSpec defines the desired state of NemoEvaluator
+type NemoEvaluatorSpec struct {
+	Image   Image           `json:"image,omitempty"`
+	Command []string        `json:"command,omitempty"`
+	Args    []string        `json:"args,omitempty"`
+	Env     []corev1.EnvVar `json:"env,omitempty"`
+	// The name of an secret that contains authn for the NGC NIM service API
+	Labels         map[string]string            `json:"labels,omitempty"`
+	Annotations    map[string]string            `json:"annotations,omitempty"`
+	NodeSelector   map[string]string            `json:"nodeSelector,omitempty"`
+	Tolerations    []corev1.Toleration          `json:"tolerations,omitempty"`
+	PodAffinity    *corev1.PodAffinity          `json:"podAffinity,omitempty"`
+	Resources      *corev1.ResourceRequirements `json:"resources,omitempty"`
+	Expose         Expose                       `json:"expose,omitempty"`
+	LivenessProbe  Probe                        `json:"livenessProbe,omitempty"`
+	ReadinessProbe Probe                        `json:"readinessProbe,omitempty"`
+	StartupProbe   Probe                        `json:"startupProbe,omitempty"`
+	Scale          Autoscaling                  `json:"scale,omitempty"`
+	Metrics        Metrics                      `json:"metrics,omitempty"`
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:default:=1
+	Replicas     int    `json:"replicas,omitempty"`
+	UserID       *int64 `json:"userID,omitempty"`
+	GroupID      *int64 `json:"groupID,omitempty"`
+	RuntimeClass string `json:"runtimeClass,omitempty"`
+
+	Mongodb       Mongodb       `json:"mongodb"`
```

Comment:
These required params are going to cause a lot of pain when microservices switch to other types of DBs etc. For e.g. say version 1 of the service uses MongoDB and version 2 switches to PostgreSQL, in the version 2 CR MongoDB becomes an optional field, but the CRD says its required (otherwise existing version 1 CRs will fail).

Best way to handle this is to make all these optional in the API and enforce in the controller logic on required dependencies per microservice version and propagate as a CR status/conditions. Even better if we can add a validation webhook to verify these.

Reply:
Makes sense. Fixed it.